### PR TITLE
Allow trailing bits while decoding base64

### PR DESCRIPTION
**Pull request by moreal**
## Description

This pull request resolves https://github.com/RustPython/RustPython/issues/4336.

Those lines to throw the `base64::DecodeError::InvalidLastSymbol{ .. }` are the below lines.

```rust
// https://github.com/marshallpierce/rust-base64/blob/10b73f67c4b19dc7fc1f4c4ae250558cba90c8d8/src/engine/fast_portable/decode_suffix.rs#L141-L147
    if !decode_allow_trailing_bits && (leftover_bits & mask) != 0 {
        // last morsel is at `morsels_in_leftover` - 1
        return Err(DecodeError::InvalidLastSymbol(
            start_of_leftovers + morsels_in_leftover - 1,
            last_symbol,
        ));
    }
```

And it provides an option, `decode_allow_trailing_bits`, to pass the check. This pull request uses the config option.

## Test

You can copy the below line and run the RustPython build on the `main` branch and on this pull request.

```python
import base64; base64.b64decode(b'3d==')
```

### CODE REVIEW

**Comment by DimitrisJim** [comment]
Surprised nothing tested for this, we should add a test for this in `extra_tests/snippets/stdlib_binascii`.

**Comment by moreal** [comment]
> Surprised nothing tested for this, we should add a test for this in `extra_tests/snippets/stdlib_binascii`.

I added the example (i.e., `3d==`) in `extra_tests/snippets/stdlib_binascii` at [d5b9664](https://github.com/RustPython/RustPython/pull/4344/commits/d5b96642717575f3bfe6a135cc71f64e72843f1d) commit. Is it enough?

**Comment by fanninpm** [comment]
> Surprised nothing tested for this

It might be a good idea to update `test_binascii.py`.

**Comment by moreal** [comment]
> It might be a good idea to update `test_binascii.py`.

To check I understood well, do you mean:

 - Add testcase to `test_binascii.py` in python/cpython
 - Add testcase to `test_binascii.py` in RustPython/RustPython
 - Update `test_binascii.py` to latest (i.e., 3.11 `test_binascii.py`)

**Comment by fanninpm** [comment]
I'm sorry if I wasn't clear. I meant "update `Lib/test/test_binascii.py` to CPython 3.11". That version has the test case that covers what you're adding here (specifically, `test_base64_strict_mode()`).

**Comment by moreal** [comment]
I updated `Lib/test/test_binascii.py`. Is there someone to review this pull request?

File: Lib/test/test_binascii.py [hunk 1]
```diff
@@ -117,6 +114,51 @@ def addnoise(line):
         # empty strings. TBD: shouldn't it raise an exception instead ?
         self.assertEqual(binascii.a2b_base64(self.type2test(fillers)), b'')
 
+    # TODO: RUSTPYTHON
+    @unittest.expectedFailure
```
Comment (fanninpm) [inline Lib/test/test_binascii.py:118]: Please break these lines out into a separate commit.

File: Lib/test/test_binascii.py [hunk 2]
```diff
@@ -368,8 +384,6 @@ def test_qp(self):
         self.assertEqual(b2a_qp(type2test(b'a.\n')), b'a.\n')
         self.assertEqual(b2a_qp(type2test(b'.a')[:-1]), b'=2E')
 
-    # TODO: RUSTPYTHON
-    @unittest.expectedFailure
```
Comment (fanninpm) [inline Lib/test/test_binascii.py:372]: Please break these lines out into a separate commit.

**Review by fanninpm — CHANGES_REQUESTED** [review]
Just a few little things, and then this will be good to go. Generally, we have one commit to update the CPython source, and an _additional_ commit for changes specific to RustPython.

File: Lib/test/test_binascii.py (hunk 1 above)
Comment (moreal) [reply Lib/test/test_binascii.py:118]: I separated these lines in https://github.com/RustPython/RustPython/pull/4344/commits/04537b1993035f78aaab94b12edc31e03d10882e commit. 🙏🏻

File: Lib/test/test_binascii.py (hunk 2 above)
Comment (moreal) [reply Lib/test/test_binascii.py:372]: I separated these lines in https://github.com/RustPython/RustPython/pull/4344/commits/dc9e958bb3172416e6d3c7747f32778c02436e13 commit. 🙏🏻

**Comment by fanninpm** [comment]
Out of curiosity, what part of `test_base64_strict_mode()` is causing it to fail?

**Comment by moreal** [comment]
> Out of curiosity, what part of `test_base64_strict_mode()` is causing it to fail?

Since CPython 3.11, A new keyword parameter `strict_mode` was added to `binascii.a2b_base64`. [^1] And we should implement it in `stdlib/src/binascii.rs#decl::a2b_binascii`. But I opened this pull request to resolve https://github.com/RustPython/RustPython/issues/4336 so I marked it with `unittest.expectedFailure`.

[^1]: https://docs.python.org/3/library/binascii.html?highlight=binascii#binascii.a2b_base64

**Comment by moreal** [comment]
I opened a new issue https://github.com/RustPython/RustPython/issues/4374 for the `strict_mode` feature task.

**Review by fanninpm — APPROVED** [review]

**Comment by fanninpm** [comment]
@DimitrisJim LGTM
